autoextract_poet/page_inputs.py

```diff
@@ -0,0 +1,59 @@
+from typing import ClassVar, List, Optional, Type
+
+import attr
+
+from autoextract_poet.items import (
+    Article,
+    Item,
+    Product,
+)
+
+
+@attr.s(auto_attribs=True)
+class ResponseData:
```

Comment:
I think generics can be leveraged here to have better type checking. Example:
```py
from typing import TypeVar, Generic, ClassVar, Type


class Item:
    pass

class Product(Item):
    pass

T = TypeVar('T', bound=Item)

class ResponseData(Generic[T]):
    item_class: ClassVar[Type[T]]

    def to_items(self) -> T:
        return self.item_class()

class ProductResponseData(ResponseData[Product]):
    item_class = Product

print(type(ProductResponseData().to_items()))
```

Reply:
@victor-torres what about this proposal?

Reply:
I might have missed something here but in my tests, it didn't present any advantage over a simple type annotation such as `Type["Item"]`.

PyCharm was still identifying return type as `Item` instead of `Product` (an `Item` subclass) when the `to_items` method was being called from an instance of a `Product` class.

What happens in your test is that you're dynamically checking the type while mypy, PyCharm, and other solutions perform a static type check. It's not the same thing. I think @ejulio's example followed the same approach of yours.

Reply:
mypy is able to perform better checking. `mypy` fails with:
```
error: "Item" has no attribute "product_method"
```
on this code:
```py
class Item:
    pass

class Product(Item):
    def product_method(self):
        pass

class ResponseData():
    item_class: ClassVar[Type[Item]]

    def to_items(self) -> Item:
        return self.item_class()

class ProductResponseData(ResponseData):
    item_class = Product

p = ProductResponseData()
item = p.to_items()
item.product_method()
```

But don't fail with this code:
```py
class Item:
    pass

class Product(Item):
    def product_method(self):
        pass

T = TypeVar('T', bound=Item)

class ResponseData(Generic[T]):
    item_class: ClassVar[Type[T]]

    def to_items(self) -> T:
        return self.item_class()

class ProductResponseData(ResponseData[Product]):
    item_class = Product

p = ProductResponseData()
item = p.to_items()
item.product_method()
```

Also, PyCharm is able to suggest me `product_method` for item in the later but not in the first case.

Reply:
I got the idea from https://mypy.readthedocs.io/en/stable/generics.html?highlight=generics

Reply:
> Also, PyCharm is able to suggest me `product_method` for item in the later but not in the first case.

It will work because we're talking about a class attribute specific to the `Product` class. 

The problem I'm trying to describe here is that AFAIK there's no way to subclass the base class without having to override those class methods to annotate them with the proper return type.

We're also probably mixing two different contexts here: Page Inputs and Items.

- https://stackoverflow.com/questions/39205527/can-you-annotate-return-type-when-value-is-instance-of-cls/39205612#39205612
- https://github.com/python/typing/issues/58
- https://github.com/python/mypy/issues/1212
- https://github.com/python/mypy/issues/3661

Reply:
In your example above, PyCharm still shows the return type of `ProductResponseData.to_item()` as being `T` and not `Product` as we would expect.

Reply:
@victor-torres PyCharm correctly infers that item is a Product in @ivanprado's `item = p.to_items()` example, even if it shows that `to_items()` returns T. In the first @ivanprado's example PyCharm shows a warning. I guess it is the same with mypy. Could you please check it again?

Reply:
This is kind-of important, because without proper typing support we loose a many of advantages of having these Item classes.

Reply:
> @victor-torres PyCharm correctly infers that item is a Product in @ivanprado's item = p.to_items() example, even if it shows that to_items() returns T.

That's true. I was looking at the return type of the `to_item` function only.

I've just implemented @ivanprado's suggestion.